contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,193 @@
+not being compatible with the existing `kubectl config set-credentials` implementation.
+
+```go
+const (
+    AuthProviderExecKindBearerToken = "bearer-token"
+    AuthProviderExecKindClientCert  = "tls-certificate"
+)
+
+type AuthProviderConfig struct {
+    Name   string            `json:"name"`
+    Config map[string]string `json:"config"`
+
+    // Exec should only be provided if Name and Config aren't.
+    Exec ExecAuthProviderConfig `json:"exec"`
+}
+
+type ExecAuthProviderConfig struct {
+    Command string   `json:"command"`
+    Args    []string `json:"args"`
+    // Env defines additional environment variables to expose to the process. These
+    // are unioned with the host's environment, with variable values lists here having
+    // precedence.
+    Env []ExecEnvVar `json:"env"`
+    // Kind determines the kind of credentials the command will output to stdout.
+    // This can be "bearer-token" or "tls-certificate".
+    Kind string `json:"kind"`
+}
+
+type ExecEnvVar struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+
+    // TODO: Load env vars from files or from other envs?
+}
+```
+
+This would allow a user block of a kubeconfig to declare the following:
+
+```yaml
+users:
+- name: mmosley
+  user:
+    auth-provider:
+      exec:
+        kind: bearer-token
+        command: /bin/kubectl-login
+        args: ["hello", "world"]
+        env:
+        - name: "VERSION"
+          value: "v1.0.1"
+```
+
+The AWS authenticator above example becomes:
+
+```
+users:
+- name: kubernetes-admin
+  user:
+    auth-provider:
+      exec:
+        kind: bearer-token
+        command: heptio-authenticator-aws
+        # CLUSTER_ID and ROLE_ARN should be replaced with actual desired values.
+        args: ["token", "-i", "(CLUSTER_ID)", "-r", "(ROLE_ARN)"]
+```
+
+## TLS client certificate support
+
+The current The auth provider interface doesn't let the user modify the dialer,
+only wrap the transport.
+
+```
+type AuthProvider interface {
+    // WrapTransport allows the plugin to create a modified RoundTripper that
+    // attaches authorization headers (or other info) to requests.
+    WrapTransport(http.RoundTripper) http.RoundTripper
+    // Login allows the plugin to initialize its configuration. It must not
+    // require direct user interaction.
+    Login() error
+}
+```
+
+Since this doesn't let a `AuthProvider` supply things like client certificates,
+the signature of the `AuthProvider` should change too ([with corresponding changes
+to `k8s.io/client-go-transport`](
+https://gist.github.com/ericchiang/7f5804403b359ebdf79dcf76c4071bff)):
+
+```
+import (
+    "k8s.io/client-go/transport"
+    // ...
+)
+
+type AuthProvider interface {
+    // UpdateTransportConfig updates a config by adding a transport wrapper,
+    // setting a bearer token (should ignore if one is already set), or adding
+    // TLS client certificate credentials.
+    UpdateTransportConfig(c *transport.Config)
```

Comment:
is this called just once at process start?

Reply:
Yes, but it can still instrument or wrap the transport. Will add a comment.